google/cloud/bigtable/data/_async/client.py

```diff
@@ -755,6 +755,10 @@ async def read_rows_with_semaphore(query):
             async with concurrency_sem:
                 # calculate new timeout based on time left in overall operation
                 shard_timeout = next(rpc_timeout_generator)
+                if shard_timeout <= 0:
+                    raise DeadlineExceeded(
+                        "Operation timeout exceeded before starting query"
```

Comment:
maybe before starting subquery?

Reply:
This is the earliest we can check after getting through the semaphore. 

There is a similar check at the very beginning as part of `_get_timeouts` though (which raises an `AttributeError`)